wrapanapi/containers/providers/rhopenshift.py

```diff
@@ -0,0 +1,874 @@
+            entities.append(entity)
+        return entities
+
+    def list_docker_registry(self):
+        """Returns list of docker registries"""
+        entities = []
+        entities_j = self.o_api.get('imagestream')[1]['items']
+        for entity_j in entities_j:
+            if 'dockerImageRepository' not in entity_j['status']:
+                continue
+            meta = entity_j['metadata']
+            entity = ImageRegistry(self, meta['name'],
+                                   entity_j['status']['dockerImageRepository'],
+                                   meta['namespace'])
+            if entity not in entities:
+                entities.append(entity)
+        return entities
+
+    def list_project(self):
+        """Returns list of projects"""
+        entities = []
+        entities_j = self.o_api.get('project')[1]['items']
+        for entity_j in entities_j:
+            meta = entity_j['metadata']
+            entity = Project(self, meta['name'])
+            entities.append(entity)
+        return entities
+
+    def list_template(self):
+        """Returns list of templates"""
+        entities = []
+        entities_j = self.o_api.get('template')[1]['items']
+        for entity_j in entities_j:
+            meta = entity_j['metadata']
+            entity = Template(self, meta['name'], meta['namespace'])
+            entities.append(entity)
+        return entities
+
+    def list_docker_image(self):
+        """Returns list of images (Docker registry only)"""
+        entities = []
+        entities_j = self.o_api.get('image')[1]['items']
+        for entity_j in entities_j:
+            if 'dockerImageReference' not in entity_j:
+                continue
+            _, name, image_id, _ = Image.parse_docker_image_info(entity_j['dockerImageReference'])
+            entities.append(Image(self, name, image_id))
+        return entities
+
+    def list_deployment_config(self):
+        """Returns list of deployment configs"""
+        entities = []
+        entities_j = self.o_api.get('deploymentconfig')[1]['items']
+        for entity_j in entities_j:
+            meta = entity_j['metadata']
+            entity = DeploymentConfig(self, meta['name'], meta['namespace'])
+            entities.append(entity)
+        return entities
+
+    def deploy_template(self, template, tags=None, password='smartvm'):
+        """Deploy a VM from a template
+
+        Args:
+            template: (str) The name of the template to deploy
+            tags: (dict) dict with tags if some tag isn't passed it is set to 'latest'
+            password: this password will be set as default everywhere
+        Returns: dict with parameters necessary for appliance setup or None if deployment failed
+        """
+        self.logger.info("starting template {t} deployment".format(t=template))
+        self.does_template_exist(namespace=self.default_namespace, name=template)
+        password = password
```

Comment:
**REQUIRED**: These two lines ^ look a bit odd, I would expect line 169 to `if` and I am quite not sure what line 170 is trying to achieve :)

Reply:
170 - removed. this was my habbit to always create local variables for parameters :)
169 - will raise TimeoutError exception if template doesn't exist. Each operation thru openshift api may take some time. so, All does_*_exist methods are implemented using wait_for